File: model/frames.py
"""Frame conversions for the dead-reckoning pipeline.

Local ENU tangent plane: origin is the first valid GPS fix. East/North in
metres. Equirectangular projection — accurate at the sub-km / few-km scales
we care about for a car demo.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

EARTH_RADIUS_M = 6_378_137.0


@dataclass(frozen=True)
class ENUOrigin:
    lat_deg: float
    lon_deg: float

    @property
    def cos_lat(self) -> float:
        return math.cos(math.radians(self.lat_deg))


def latlon_to_en(lat_deg: float, lon_deg: float, origin: ENUOrigin) -> tuple[float, float]:
    """Equirectangular projection to local East/North metres."""
    d_lat = math.radians(lat_deg - origin.lat_deg)
    d_lon = math.radians(lon_deg - origin.lon_deg)
    east = EARTH_RADIUS_M * d_lon * origin.cos_lat
    north = EARTH_RADIUS_M * d_lat
    return east, north


def en_to_latlon(east_m: float, north_m: float, origin: ENUOrigin) -> tuple[float, float]:
    d_lat = north_m / EARTH_RADIUS_M
    d_lon = east_m / (EARTH_RADIUS_M * origin.cos_lat)
    return origin.lat_deg + math.degrees(d_lat), origin.lon_deg + math.degrees(d_lon)


def integrate_heading(heading_rad: float, gyro_z: float, dt: float) -> float:
    """Yaw-only integration. Assumes vehicle roughly level — fine for cars."""
    return _wrap_pi(heading_rad + gyro_z * dt)


def device_accel_to_world(accel_x: float, accel_y: float, heading_rad: float) -> tuple[float, float]:
    """Rotate horizontal device-frame acceleration to world-frame (E, N).

    heading_rad is measured East-of-North (standard nav convention).
    """
    c, s = math.cos(heading_rad), math.sin(heading_rad)
    east = accel_x * c - accel_y * s
    north = accel_x * s + accel_y * c
    return east, north


def _wrap_pi(a: float) -> float:
    return (a + math.pi) % (2 * math.pi) - math.pi

File: web/src/kalman/frames.ts
/**
 * Frame conversions. Ports of model/frames.py — see that file for the design.
 * Kept deliberately literal so the TS + Python filters produce matching output
 * on the same input.
 */

export const EARTH_RADIUS_M = 6_378_137;

export interface ENUOrigin {
  latDeg: number;
  lonDeg: number;
  cosLat: number;
}

export function makeOrigin(latDeg: number, lonDeg: number): ENUOrigin {
  return { latDeg, lonDeg, cosLat: Math.cos((latDeg * Math.PI) / 180) };
}

/** Equirectangular projection to local East/North metres. */
export function latLonToEN(latDeg: number, lonDeg: number, origin: ENUOrigin): [number, number] {
  const dLat = ((latDeg - origin.latDeg) * Math.PI) / 180;
  const dLon = ((lonDeg - origin.lonDeg) * Math.PI) / 180;
  return [EARTH_RADIUS_M * dLon * origin.cosLat, EARTH_RADIUS_M * dLat];
}

export function enToLatLon(eastM: number, northM: number, origin: ENUOrigin): [number, number] {
  const dLat = northM / EARTH_RADIUS_M;
  const dLon = eastM / (EARTH_RADIUS_M * origin.cosLat);
  return [origin.latDeg + (dLat * 180) / Math.PI, origin.lonDeg + (dLon * 180) / Math.PI];
}

/** Yaw-only integration. Vehicle roughly level. */
export function integrateHeading(headingRad: number, gyroZ: number, dt: number): number {
  return wrapPi(headingRad + gyroZ * dt);
}

/**
 * Rotate horizontal device-frame acceleration to world-frame (E, N).
 * Body frame: +x forward, +y right. Heading = East-of-North (compass).
 *   forward_world = (sin θ,  cos θ);  right_world = (cos θ, -sin θ).
 */
export function deviceAccelToWorld(
  accelX: number,
  accelY: number,
  headingRad: number,
): [number, number] {
  const c = Math.cos(headingRad);
  const s = Math.sin(headingRad);
  return [accelX * s + accelY * c, accelX * c - accelY * s];
}

function wrapPi(a: number): number {
  return ((a + Math.PI) % (2 * Math.PI)) - Math.PI;
}
